File: Programmers/[86052] 빛의 경로 사이클/SUbbb/86052.java
import java.util.*;

class Solution {
    private int m;
    private int n;
    private boolean[][][] visited;
    private String[] grid;
    // 0: 아래, 1: 왼, 2: 위, 3: 오른
    private int[][] dir = {{-1, 0}, {0, -1}, {1, 0}, {0, 1}};
        
    public int[] solution(String[] grid) {
        m = grid.length;
        n = grid[0].length();
        visited = new boolean[m][n][4];
        this.grid = grid;
        
        ArrayList<Integer> answer = new ArrayList<>();
        
        // 모든 격자에서 4방향으로 다 방문
        for (int i = 0; i < m; i++)
            for (int j = 0; j < n; j++)
                for (int d = 0; d < 4; d++)
                    if (!visited[i][j][d])
                        answer.add(go(i, j, d));
        
        return answer.stream().sorted().mapToInt(i -> i).toArray();
    }
    
    private int go(int x, int y, int d) {
        // 이동 거리
        int cnt = 0;
 
        while (!visited[x][y][d]) {
            cnt++;
            visited[x][y][d] = true;
 
            if (grid[x].charAt(y) == 'L')
                d = d == 0 ? 3 : d - 1; // 좌회전
            else if (grid[x].charAt(y) == 'R')
                d = d == 3 ? 0 : d + 1; // 우회전
            
            // 다음 좌표 계산
            x = (x + dir[d][0] + m) % m;
            y = (y + dir[d][1] + n) % n;
        }
 
        return cnt;
    }
}

File: Programmers/[86052] 빛의 경로 사이클/JFe/86052.py
from collections import deque

dx = [1, 0, -1, 0]
dy = [0, 1, 0, -1]

def check(grid, visited, x, y, d):
    r, c, cnt = len(grid), len(grid[0]), 0
    q = deque()
    q.append([x, y, d])
    # 빛 이동
    while q:
        cur_x, cur_y, cur_d = q.popleft()
        # 이미 방문했으면 break
        if visited[cur_y][cur_x][cur_d]: break
        visited[cur_y][cur_x][cur_d] = True
        # 다음 위치 계산 (격자 끝을 넘어가면 반대쪽 끝으로 가도록)
        next_x, next_y = cur_x + dx[cur_d], cur_y + dy[cur_d]
        if next_x == c: next_x = 0
        elif next_x == -1: next_x = c - 1
        if next_y == r: next_y = 0
        elif next_y == -1: next_y = r - 1
        # 다음 방향 계산
        if grid[next_y][next_x] == 'S': next_d = cur_d
        elif grid[next_y][next_x] == 'L': 
            if cur_d == 0: next_d = 3
            else: next_d = cur_d - 1
        elif grid[next_y][next_x] == 'R': 
            if cur_d == 3: next_d = 0
            else: next_d = cur_d + 1
        q.append([next_x, next_y, next_d])
        cnt += 1
    return cnt
        
def solution(grid):
    answer = []
    r, c = len(grid), len(grid[0])
    visited = [[[False for _ in range(4)]for _ in range(c)]for _ in range(r)]
    # 시작 지점, 방향 바꿔가면서 사이클 체크
    for i in range(r):
        for j in range(c):
            for k in range(4):
                result = check(grid, visited, j, i, k)
                if result != 0: answer.append(result)
    return sorted(answer)